framework/src/task_manager/task_manager_core.c

```diff
@@ -663,59 +671,171 @@ int taskmgr_get_handle_by_pid(int pid)
+{
+	if (tm_broadcast_msg[msg - 1] == msg) {
+		return OK;
+	}
+	return TM_UNREGISTERED_MSG;
+}
+
+static tm_broadcast_info_t *taskmgr_search_broadcast_info(int msg, int handle)
+{
+	tm_broadcast_info_t *curr;
+
+	if (taskmgr_check_broad_msg(msg) != OK) {
+		tmdbg("Not regiseterd broadcast msg: %d\n", msg);
+		return NULL;
+	}
+	curr = (tm_broadcast_info_t *)sq_peek(&TM_BROADCAST_INFO_LIST(handle));
+	while (curr) {
+		if (curr->msg == msg) {
+			return curr;
+		}
+		curr = sq_next(curr);
+	}
+	return curr;
+}
+
 static void taskmgr_broadcast(int msg)
 {
 	int handle;
 	int fd;
 	int ret;
 	union sigval msg_broad;
+	tm_broadcast_info_t *broadcast_info;
 
 	fd = taskmgr_get_drvfd();
 	if (fd < 0) {
 		return;
 	}
-	msg_broad.sival_int = msg;
+
+	ret = taskmgr_check_broad_msg(msg);
+	if (ret == TM_UNREGISTERED_MSG) {
+		return;
+	}
+
 	for (handle = 0; handle < CONFIG_TASK_MANAGER_MAX_TASKS; handle++) {
 		if (TM_LIST_ADDR(handle) != NULL) {
 			ret = taskmgr_get_task_state(handle);
 			if (ret == TM_APP_STATE_STOP || ret == TM_APP_STATE_UNREGISTERED) {
 				continue;
 			}
-			ret = taskmgr_check_msg_mask(msg, handle);
+			ret = ioctl(fd, TMIOC_BROADCAST, TM_PID(handle));
 			if (ret != OK) {
 				continue;
 			}
-			ret = ioctl(fd, TMIOC_BROADCAST, TM_PID(handle));
-			if (ret != OK) {
+			broadcast_info = taskmgr_search_broadcast_info(msg, handle);
+			if (broadcast_info == NULL) {
 				continue;
 			}
+			msg_broad.sival_ptr = broadcast_info;
 			(void)sigqueue(TM_PID(handle), SIGTM_BROADCAST, msg_broad);
 		}
 	}
 }
 
+static void taskmgr_broadcast_msg_init(void)
+{
+	int chk_idx;
+	int msg_num;
+
+	for (chk_idx = 0; chk_idx < TM_BROADCAST_MSG_MAX + CONFIG_TASK_MANAGER_MAX_TASKS; chk_idx++) {
+		msg_num = chk_idx + 1;
+		if (chk_idx < (TM_BROADCAST_MSG_MAX - 1)) {
+			tm_broadcast_msg[chk_idx] = msg_num;
+		} else {
+			tm_broadcast_msg[chk_idx] = -1;
+		}
+	}
+}
+
+static int taskmgr_is_tm_broadcast_msg_init(void)
+{
+	if ((tm_broadcast_msg[0] != 1) && (tm_broadcast_msg[0] != -1)) {
+			return 0;
+	}
+	return 1;
+}
+
 static int taskmgr_set_msg_cb(int type, void *data, int pid)
 {
 	int handle;
+	int ret;
+	tm_broadcast_info_t *broadcast_info;
 
 	if (data == NULL) {
 		return TM_INVALID_PARAM;
 	}
-
 	handle = taskmgr_get_handle_by_pid(pid);
 	if (handle == TM_UNREGISTERED_APP) {
 		return handle;
 	}
-
 	if (type == TYPE_UNICAST) {
 		TM_UNICAST_CB(handle) = (_tm_unicast_t)data;
 	} else {
-		TM_MSG_MASK(handle) = ((tm_broadcast_t *)data)->msg_mask;
-		TM_BROADCAST_CB(handle) = ((tm_broadcast_t *)data)->cb;
+		if (!taskmgr_is_tm_broadcast_msg_init()) {
+			taskmgr_broadcast_msg_init();
+		}
+		ret = taskmgr_check_broad_msg(((tm_broadcast_info_t *)data)->msg);
+		if (ret == TM_UNREGISTERED_MSG) {
+			return ret;
+		}
+		broadcast_info = taskmgr_search_broadcast_info(((tm_broadcast_info_t *)data)->msg, handle);
+		if (broadcast_info == NULL) {
+			broadcast_info = (tm_broadcast_info_t *)TM_ALLOC(sizeof(tm_broadcast_info_t));
```

Comment:
you'd better check the fail of TM_ALLOC case

Reply:
Done